### PR TITLE
PUBDEV-8416: Add stopping parameters to SE

### PR DESCRIPTION
**Pull request by tomasfryda**
https://h2oai.atlassian.net/browse/PUBDEV-8416

Up until this commit, users could provide stopping parameters only using `metalearner_params`. This commit exposes them as StackedEnsemble's parameters. 

To make this change backward compatible the code first checks if the parameter has a default value in `metalearner_params` and if it does it will overwrite it using the value that is in Stacked Ensemble's parameters.

In other words, parameters set in `metalearner_params` have higher priority and will be used if specified. 

The problem with this approach is that it is not consistent with what we currently have in SE, e.g.,`weights_column` will always be used from SE's params and not from the `metalearner_params`. I can make the changes both ways to make SE parameter setting consistent. 

Any opinions @sebhrusen @ledell ?

### CODE REVIEW

File: h2o-py/h2o/estimators/stackedensemble.py [hunk 1]
```diff
@@ -155,10 +159,27 @@ def __init__(self,
                random number)
                Defaults to ``-1``.
         :type seed: int
+        :param score_each_iteration: Whether to score during each iteration of model training.
```
Comment (michalkurka) [inline h2o-py/h2o/estimators/stackedensemble.py:162]: What about score_tree_interval for tree-based models?

File: h2o-algos/src/main/java/hex/ensemble/Metalearner.java [hunk 2]
```diff
@@ -116,6 +132,12 @@ protected void setCommonParams(P parms) {
     parms._max_runtime_secs = _maxRuntimeSecs;
     parms._weights_column = _model._parms._weights_column;
     parms._offset_column = _model._parms._offset_column;
+
+    // Checking if it was set in "metalearner_params" to keep backward compatibility
```
Comment (michalkurka) [inline h2o-algos/src/main/java/hex/ensemble/Metalearner.java:136]: instead of using reflection it would be better to introduce an API on model Parameters for setting early stopping criteria

in tree-based you have score_tree_interval, in GLM score_iteration_interval 

we could eg. add a method setScoreIterationInteval to Model.Parameters that would set the parameter if it is suppored

File: h2o-py/h2o/estimators/stackedensemble.py (hunk 1 above)
Comment (tomasfryda) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: Thanks, I forgot about that one.

It makes me wonder if this PR is a good idea. We can already set the parameters via `metalearner_params` and exposing parameters that are applicable only on some meta learners on the SE level doesn't seem like a good idea to me.

Comment (sebhrusen) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: @tomasfryda I thought the ticket was only about `stopping_*` params that define the early stopping feature and that are supported by all metalearners I think.
I'm much less convinced about `score_each_iteration` or `score_tree_interval` that are not really necessary to use early stopping, just optional additions.

Comment (tomasfryda) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: Ok, I thought it should include even these as I think they influence the stopping - to be able to early stop we have to have scored iterations/trees, in other words, it defines the "resolution" used by the early stopping. At least this is how it worked in GLM last time I looked.

I will remove the `score_each_iteration`.

Comment (michalkurka) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: @sebhrusen Without those parameters, the early stopping feature is not reproducible and I would consider it problematic in general. If you don't know how often H2O scores then it is very difficult to correctly set stopping rounds.

Comment (sebhrusen) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: @michalkurka I get your point about reproducibility. My concern is purely semantic, and deals with API choices.

To me the problem is similar to the decision that was made long ago to expose only the `stopping_*` params in 
AutoML: I guess that only those ones have been exposed to AutoML because they are enough to define the early stopping feature. The others depend on each model trained internally. If the user wanted to specify the optional ones for each algo, now s·he could override those using `algo_parameters`. Not ideal, but otherwise we could as well expose any possible algo param as a first-class param.

For SE, I think it's similar: the `stopping_*` are enough to define the feature—a feature otherwise made available to all algos, except SE until now—, but if the user needs to fine tune, s·he will have to consider the specific metalearner used by the SE, and may as well set those params as `metalearner_params`.

Comment (sebhrusen) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: @ledell wdyt?

Comment (michalkurka) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: I think the difference between AutoML and SE is that AutoML by default has predictable scoring intervals and users can reason about it and use early stopping to their advantage.

I am thinking that these parameters are not needed on SE level because without understanding what metalearner is used and what parameters goes in the ability to set the early stopping parameters is limited anyway.

Comment (ledell) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: @sebhrusen @tomasfryda Let's chat on Slack to discuss in more detail and then return to the thread here...

Comment (sebhrusen) [reply h2o-py/h2o/estimators/stackedensemble.py:162]: @michalkurka we agreed on not to expose the params at all, and as a result @tomasfryda will close this PR.
It was based on a misconception that some of the high-level params were already forwarded to the metalearner, but this is actually not the case: only data/input-related params are, not algo-related ones.

Thanks for being watchful on this.

**Comment by ledell** [comment]
After a lengthy discussion, we decided it was cleaner to keep metalearner related params in the `metalearner_params` arg and not expose/duplicate them at the top level.  Instead we will modify the JIRA to tackle the actual issue which is that early stopping params from AutoML are not getting propagated down to the Stacked Ensemble (we will instead pass them through the `metalearner_params` arg directly).
